Remove pytz dependency, update minimum python version to 3.9, extend tests

do3cc opened the pull request:
This is the PR to implement #42 
I've reviewed timezone/utils.py and noticed a lot of code isn´t used at all.
Then I added tests for windows and macos to make sure, nothing breaks.
Since nothing broke, I got worried and extended the tests to use timezones as provided by Python 3.9.
Then my windows tests failed and I added the necessary dependency for windows.
Now that I am writing this, I am wondering whether we should depend on tzinfo, for windows, as we don´t use the available timezones anywhere. However, I like that there are tests for that.

justinmayer commented:
Thanks for your work on this, Patrick.

@venthur / @hugovk: I am preparing for a trip and could use an extra pair of eyes. Would you mind taking a look and providing any feedback you may have?

do3cc commented:
After this I also downloaded pelican and made a test run with my modified branch. No failures. However I saw a few skipped tests due to missing locales.

venthur commented on .github/workflows/main.yml:
```diff
@@ -11,11 +11,16 @@ permissions:
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        python-version: ["3.9", "3.13"]
+        include:
+          - { os: "ubuntu-latest", python-version: "3.10" }
+          - { os: "ubuntu-latest", python-version: "3.11" }
+          - { os: "ubuntu-latest", python-version: "3.12" }
```
I appreciate this change very much, I do the same thing with my packages. I just wonder, why do you limit the python-version to two versions and explicitly add the remaining ones with ubuntu only?

```suggestion
        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13"]
```

do3cc replied:
Premature optimization? My gut feeling tells me that operating specific problems are likely to happen with either all python versions, only late or only new python versions and not in between.
But it is a tradeoff between having Microsoft spend more money on Open Source or a simpler config :-)

venthur commented on tox.ini:
```diff
@@ -1,9 +1,8 @@
-    py3.8: python3.8
     py3.9: python3.9
     py3.10: python3.10
     py3.11: python3.11
```
Shouldn't 3.12 and 3.13 be mentioned here too?

do3cc replied:
It seems I completely forgot about tox.ini or might have accidentally forgot to add my changes to it. I will check at home.
If you want, I can also spend a bit of time to migrate tox.ini into pyproject.toml. I think it is a tiny bit more verbose but would mean one less config file.
I initially hesitated to do, because maybe the project prefers to have the config separated.

venthur replied:
I'd suggest to only fix tox.ini here to reflect the other changes. We could have the migration to pyproject.toml a new pull request, since it is unrelated to the tz issue you're fixing.

venthur commented on tests/test_feedgenerator.py:
```diff
@@ -29,14 +30,14 @@
     Chinese: 老师是四十四，是不是？
     Finnish: Mustan kissan paksut posket. (ah, no special chars) Kärpänen sanoi kärpäselle: tuu kattoon kattoon ku kaveri tapettiin tapettiin.
-    </description><language>en</language><lastBuildDate>%DATE%</lastBuildDate><item><title>Hello</title><link>http://www.holovaty.com/t%C3%A4st/</link><description>Testing.</description><pubDate>Thu, 11 Aug 2016 00:00:00 -0000</pubDate></item></channel></rss>"""
+    </description><language>en</language><lastBuildDate>%DATE%</lastBuildDate><item><title>Hello</title><link>http://www.holovaty.com/t%C3%A4st/</link><description>Testing.</description><pubDate>Thu, 11 Aug 2016 00:00:00 -0400</pubDate></item></channel></rss>"""
```
This test case is already quite convoluted can you add a new one that focuses only on the time zone tests and keeps the other strings to a minimum?

do3cc replied:
reverted and created a dedicated test for proper timezone handling

venthur commented on tox.ini:
```diff
@@ -1,9 +1,8 @@
 [tox]
-envlist = py{3.8,3.9,3.10,3.11,3.12}
+envlist = py{3.9,3.10,3.11,3.12}
```
I took the liberty to fix that one myself.

hugovk commented on .github/workflows/main.yml:
```diff
@@ -11,11 +11,12 @@ permissions:
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
```
It can be _slightly_ quicker to start the slower ones first:
```suggestion
        os: ["windows-latest", "macos-latest", "ubuntu-latest"]
```

hugovk commented on .github/workflows/main.yml:
```diff
@@ -11,11 +11,12 @@ permissions:
   test:
-    name: Test - Python ${{ matrix.python-version }}
-    runs-on: ubuntu-latest
+    name: Test - Python ${{ matrix.python-version }} on ${{ matrix.os }}
```
Now:
<img width="264" alt="image" src="https://github.com/user-attachments/assets/926789df-47ff-4a83-b079-c85a99fd1c80" />


I usually omit this:
```suggestion
```

The default is quite concise:

<img width="256" alt="image" src="https://github.com/user-attachments/assets/ba572d54-a2eb-4eb4-8c32-6fa9aed4e2d7" />

https://github.com/termcolor/termcolor/actions/runs/13591158653

hugovk commented on .github/workflows/main.yml:
```diff
@@ -11,11 +11,12 @@ permissions:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13"]
```
Is PyPy of interest?

```suggestion
        python-version: ["pypy3.11", "3.9", "3.10", "3.11", "3.12", "3.13"]
```

do3cc replied:
I've added it, why not if it works.

hugovk commented on pyproject.toml:
```diff
@@ -25,9 +25,9 @@ classifiers = [
     "Topic :: Internet :: WWW/HTTP :: Dynamic Content",
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
```
Also remove:
```
    "Programming Language :: Python :: 3.8",
```
And add:
```
    "Programming Language :: Python :: 3.13",
```

hugovk reviewed (commented):
Also:
```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 5b811c0..d2d7027 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,7 +1,7 @@
 ---
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.6.0
+    rev: v5.0.0
     hooks:
       - id: check-added-large-files
       - id: check-ast
@@ -15,7 +15,7 @@ repos:
       - id: trailing-whitespace
 
   - repo: https://github.com/asottile/pyupgrade
-    rev: v3.15.2
+    rev: v3.19.1
     hooks:
       - id: pyupgrade
-        args: [--py38-plus]
+        args: [--py39-plus]

```

do3cc commented:
Everything addressed. Thanks for the review

venthur commented:
Thanks for your contribution and addressing all requests @do3cc! @justinmayer the PR looks good to go.

do3cc commented:
No worries, I just did so.

justinmayer commented on .pre-commit-config.yaml:
```diff
@@ -1,21 +1,21 @@
+      rev: v3.19.1
+      hooks:
+          - id: pyupgrade
+            args: [--py39-plus]
```
If it’s not too much to ask, could you (1) revert the whitespace changes in this file and (2) update the `pyupgrade` version to the most-recent `v3.20.0`?

do3cc commented:
I have also changed the github workflow again. I got some feedback from earlier reviews not do the optimizations that I just reverted and I somehow managed to loose them during the refactoring.
